grails-app/services/com/unifina/service/UserService.groovy

```diff
@@ -161,24 +163,24 @@ class UserService {
 		}
 	}
 
-	SecUser getUserFromUsernameAndPassword(String username, String password) {
+	SecUser getUserFromUsernameAndPassword(String username, String password) throws InvalidUsernameAndPasswordException {
 		PasswordEncoder encoder = new BCryptPasswordEncoder()
 		SecUser user = SecUser.findByUsername(username)
 		if (user == null) {
-			return null
+			throw new InvalidUsernameAndPasswordException("Invalid: "+username+", "+password)
```

Comment:
Let's not echo back the credentials! They could accidentally end up in our server logs for example. The message should be just vague standard message, such as _Invalid username or password._